reservation할 때 payment 추가 및 test code 작성

MyungHwanKim opened the pull request:
# Summary
feat : reservation할 때 payment 추가

- reservationForm에 PaymentForm 추가
- reservation과 payment 동시 처리로 cascadeType.ALL 추가
- reservation 수정
  - 예약 : reservation 반환 void -> reservationDto
  - 예약 정보 : reservation/info -> reservation/info/{guestId}
  - 예약 취소 : reservation -> reservation/{reservationId}

test : reservation할 때 payment 추가한 test code 작성

- 예약 취소 성공하는 일자 변경(2022년 -> 2222년)
- 현재 날짜 기준으로 7일 이내이므로 당분간 성공 케이스를 위해 변경

# Relevant

커밋 ID : c9b9427, 8c33bb7

# Screenshot

![reservationTest](https://user-images.githubusercontent.com/102682659/192923212-a00cf8cc-0ae6-496d-a3bf-64c03547b8b1.png)

payment 추가하여 test 진행한 김에 reservation test 전체 확인입니다.

# PR CheckList
아래의 요구사항을 충족하셨는지 확인하여 체크해주세요.

체크 방법은 [ ] 괄호 안에 x를 기입하시면 됩니다.

- [x] 작업한 내용에 대해 테스트를 진행했음.
- [x] 칸반보드에 해당 작업이 업데이트 되어있음.
- [x] commit 전, `git status`, `git pull`을 통해 반영되지 않은 코드가 없음을 확인함.

# To Reviewers
모두 화이팅입니다~~~!!

JmKanmo commented:
제가 프로젝트에 swagger 설정 적용해놓아서 나중을 위해  컨트롤러에 swagger 어노테이션도 붙여주세요
아래 처럼 작성하면 나중에 swagger페이지에 컨트롤러 API 전부 상세하게 나옵니다.

아래는 컨트롤러 작성 예시구요.
다른 작업하시는 분들도 마찬가지루요 (저포함)

아래와 같은식으로요 

```
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;

import java.security.Principal;

@Tag(name = "메인화면", description = "메인화면 관련 API")
@Controller
@RequiredArgsConstructor
@Slf4j
public class MainController {

    @Operation(summary = "메인 페이지 반환", description = "메인 페이지 반환 메서드")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "메인 페이지 반환에 성공했을 때의 응답 코드"),
            @ApiResponse(responseCode = "500", description = "특정 이유로 메인 페이지 반환에 실패했을 때의 응답 코드")
    })
    @GetMapping("/")
    public String main() {
        return "index";
    }

    @Operation(summary = "설정 페이지", description = "설정 페이지 반환 메서드")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "설정 페이지 반환에 성공했을 때의 응답 코드),
            @ApiResponse(responseCode = "500", description = "특정 이유로 설정 페이지 반환에 실패했을 때의 응답 코드")
    })
    @GetMapping("/settings")
    public String settings() {
        return "settings";
    }
}

```

JmKanmo commented:
이메일 인증은 저는 구글 찬성입니다 ㅎㅎ 
그리고 application-secret.yml 보단 application-db.yml이 좀 더 네이밍이 낫지 않을까용??

Haze-S commented:
> 이메일 인증은 저는 구글 찬성입니다 ㅎㅎ 그리고 application-secret.yml 보단 application-db.yml이 좀 더 네이밍이 낫지 않을까용??

secret 파일에 db, jwt, google mail 까지 같이 들어있어서 secret 으로 통일했는데 각각 다른 파일로 나누는 게 좋을까요? 🤔

Haze-S commented:
> 수고하셨습니당
> 
> 제가 프로젝트에 swagger 설정 적용해놓아서 나중을 위해 컨트롤러에 swagger 어노테이션도 붙여주세요 아래 처럼 작성하면 나중에 swagger페이지에 컨트롤러 API 전부 상세하게 나옵니다.
> 
> 아래는 컨트롤러 작성 예시구요. 다른 작업하시는 분들도 마찬가지루요 (저포함)
> 
> 아래와 같은식으로 부탁드릴게여~
> 
> ```
> import io.swagger.v3.oas.annotations.Operation;
> import io.swagger.v3.oas.annotations.responses.ApiResponse;
> import io.swagger.v3.oas.annotations.responses.ApiResponses;
> import io.swagger.v3.oas.annotations.tags.Tag;
> import lombok.RequiredArgsConstructor;
> import lombok.extern.slf4j.Slf4j;
> import org.springframework.stereotype.Controller;
> import org.springframework.ui.Model;
> import org.springframework.web.bind.annotation.GetMapping;
> 
> import java.security.Principal;
> 
> @Tag(name = "메인화면", description = "메인화면 관련 API")
> @Controller
> @RequiredArgsConstructor
> @Slf4j
> public class MainController {
> 
>     @Operation(summary = "메인 페이지 반환", description = "메인 페이지 반환 메서드")
>     @ApiResponses(value = {
>             @ApiResponse(responseCode = "200", description = "메인 페이지 반환에 성공했을 때의 응답 코드"),
>             @ApiResponse(responseCode = "500", description = "특정 이유로 메인 페이지 반환에 실패했을 때의 응답 코드")
>     })
>     @GetMapping("/")
>     public String main() {
>         return "index";
>     }
> 
>     @Operation(summary = "설정 페이지", description = "설정 페이지 반환 메서드")
>     @ApiResponses(value = {
>             @ApiResponse(responseCode = "200", description = "설정 페이지 반환에 성공했을 때의 응답 코드),
>             @ApiResponse(responseCode = "500", description = "특정 이유로 설정 페이지 반환에 실패했을 때의 응답 코드")
>     })
>     @GetMapping("/settings")
>     public String settings() {
>         return "settings";
>     }
> }
> ```

친절한 예시까지ㅎㅎ 감사합니다! 수정 완료했습니다!
401a8de

MyungHwanKim committed:
feat : reservation할 때 payment 추가

- reservationForm에 PaymentForm 추가
- reservation과 payment 동시 처리로 cascadeType.ALL 추가

MyungHwanKim committed:
test : reservation할 때 payment 추가한 test code 작성

- 예약 취소 성공하는 일자 변경(2022년 -> 2222년)
- 현재 날짜 기준으로 7일 이내이므로 당분간 성공 케이스를 위해 변경

JmKanmo commented:
코드 작성하시느라 수고 많으셨습니다. 
근데, 테스트 코드 ... gradle test 태스크로 실행해도 정상 통과 되시던가요..?  내 환경이 문제엿나 ㅠ 

CascadeType.ALL 이란 문법을 쓰셨는데, 저는 직접 써본적은 없고 개념만 어렴풋이 기억나는데,  특정 엔티티에 대해 해당 엔티티를  생성하거나 혹은 삭제할 때, 연관된 (ManyToOne, OntToOne)이엿나... 이런 연관된 객체도 생성 or 삭제하는걸로 알고있는데.,. (제가 잘못잇는거면 피드백 부탁드려요 ㅎㅎ) 

All의 경우에는 생성,삭제 전부 다 포함되는거라 이를 잘못 사용하거나, 실수 했을때, 막아주는게 없어서 그만큼 주의를 기울여서 조심해서 사용해야 한다네요. 아마 명환님도 이부분 유의하셔서 사용을 결심하셨으리라 생각이 드네요! 

그밖에, 다른 부분은 따로 피드백 드릴건 없네요. 다시 한번 수고 많으셨어요

MyungHwanKim commented:
> 코드 작성하시느라 수고 많으셨습니다. 근데, 테스트 코드 ... gradle test 태스크로 실행해도 정상 통과 되시던가요..? 내 환경이 문제엿나 ㅠ
> 
> CascadeType.ALL 이란 문법을 쓰셨는데, 저는 직접 써본적은 없고 개념만 어렴풋이 기억나는데, 특정 엔티티에 대해 해당 엔티티를 생성하거나 혹은 삭제할 때, 연관된 (ManyToOne, OntToOne)이엿나... 이런 연관된 객체도 생성 or 삭제하는걸로 알고있는데.,. (제가 잘못잇는거면 피드백 부탁드려요 ㅎㅎ)
> 
> All의 경우에는 생성,삭제 전부 다 포함되는거라 이를 잘못 사용하거나, 실수 했을때, 막아주는게 없어서 그만큼 주의를 기울여서 조심해서 사용해야 한다네요. 아마 명환님도 이부분 유의하셔서 사용을 결심하셨으리라 생각이 드네요!
> 
> 그밖에, 다른 부분은 따로 피드백 드릴건 없네요. 다시 한번 수고 많으셨어요

현재 pr을 올리기 전에 한 곳에서 test code가 실패할 걸로 예상되긴 합니다,,,!
LocalDateTime.now() 가 적용되어 있어서 이번에 다시 해보니 실패하는 경우가 보이더라고요!!!
그래서 실패할 수 없도록 수정하긴 해뒀습니다!

CascadeType.ALL 은 사실 reservation과 payment가 동시에 발생하니까 payment가 먼저 이루어지고 reservation이 이루어질 때 payment도 같이 저장되도록 하기 위해 사용하기 했습니다. 현재 둘의 관계는 OneToOne 관계이고 저희 프로젝트에서는 실제 결제할 일이 없다보니 우선 ALL을 사용하여 에약을 취소했을 때 결제 내역도 같이 삭제한다고 생각은 했지만, 실제 결제를 한다면 고려할 사항이 많다보니 이 방안이 아닌 다른 방안을 생각해야하지 않을까 생각했습니다 ㅎㅎ

MyungHwanKim committed:
feat : reservation할 때 guest point 변경

- 예약할 때 guest point 차감
- 예약할 때 guest point 보다 더 큰 point 사용 시 exception
- 예약 취소할 때 사용한 point 다시 guest point에 적용

MyungHwanKim committed:
test :  예약할 때 guest point 보다 더 큰 point 사용 시 exception test code 작성
